Infinite mana spawner

j-tomada opened the pull request:
Added 2 scripts, 2 prefabs, and 1 scene 

InfiniteMana.cs
-Should be added to InfiniteMana prefab
-Inherits from entity in order to get the TakeDamage method
-Overrides TakeDamage method to spawn 1 - 3 mana orbs instead of taking damage

ManaPickup.cs
-Should be added to ManaOrb prefab
-Allows orb to be able to be picked up via collision
-Upon collision with player, orb should be consumed and fill up the mana bar

InfiniteMana, ManaOrb
-Prefabs
-ManaOrb has a circle collider in order to be consumed

New Scene: Infinite Mana
-Scene to test mana spawner
-Press 'z' in order to call TakeDamage on the Spawner

j-tomada committed:
Overrided TakeDamage

Overrided TakeDamage method from unity

Whenever object takes damage, it should reset its hp back to full, therefore, not taking damage

j-tomada committed:
TakeDamage now instaciates ManaOrb prefab

Created a base for the ManaOrb Prefab

TakeDamage Method will now instanciate 1 - 3 mana orbs on the scene if it takes damage

j-tomada committed:
Added ManaPickup.cs

Will be used on the mana orb prefab

j-tomada committed:
Merge branch 'main' into InfiniteManaSpawner

j-tomada committed:
Created InfiniteManaSpawner Prefab

Created prefab for InfiniteManaSpawner.

Hold a Sprite Renderer and the InfiniteMana script

j-tomada committed:
Merge branch 'main' into InfiniteManaSpawner

j-tomada committed:
Merge branch 'main' into InfiniteManaSpawner

j-tomada committed:
Created prefabs and scene

Created prefabs for both infinite spawner and mana orb

Created a scene that can be used for testing

j-tomada committed:
added comments and descriptions

Added comments and descriptions to InfiniteMana.cs and ManaPickup.cs

j-tomada committed:
Made some changes for testing

Changed TakeDamage key to 'z' instead of spacebar

Changed it so each mana orb gives 10 mana back

Kainkun commented on Assets/Scripts/InfiniteMana.cs:
```diff
@@ -0,0 +1,49 @@
+
+    public override void TakeDamage(float baseDamage)
+    {
+        base.TakeDamage(baseDamage);
```
getting rid of this line will prevent its health from going down so that you don't need to set its health to something so high

Kainkun commented on Assets/Scripts/InfiniteMana.cs:
```diff
@@ -0,0 +1,49 @@
+        base.TakeDamage(baseDamage);
+
+        Vector2 orbPosition;
+        for (int i = 0; i < rand.Next(1, 4); ++i) //Generate 1 - 3 mana orbs
```
instead of using System's Random using [Random.Range](https://docs.unity3d.com/ScriptReference/Random.Range.html) is a bit more convenient and has more features

Kainkun commented on Assets/Scripts/InfiniteMana.cs:
```diff
@@ -0,0 +1,49 @@
+        Vector2 orbPosition;
+        for (int i = 0; i < rand.Next(1, 4); ++i) //Generate 1 - 3 mana orbs
+        {
+            orbPosition = new Vector2(spawnerCordsX + i + 3, spawnerCordsY); //Makes it so that orbs are separated from each other; Testing purposes
```
For example, Random has a [Random.insideUnitCircle](https://docs.unity3d.com/ScriptReference/Random-insideUnitCircle.html) That lets you randomly get a position inside of a circle. That would be useful for spawning the orbs

Kainkun commented on Assets/Scripts/InfiniteMana.cs:
```diff
@@ -0,0 +1,49 @@
+            orbPosition = new Vector2(spawnerCordsX + i + 3, spawnerCordsY); //Makes it so that orbs are separated from each other; Testing purposes
+            Instantiate(manaOrb, orbPosition, Quaternion.identity); //Creates a new Mana Orb prefab on the scene
+        }
+        base.health = maxHealth; //Health resets back to full 
```
This wont be necessary once you prevent it from taking damage

Kainkun commented on Assets/Scripts/InfiniteMana.cs:
```diff
@@ -0,0 +1,49 @@
+    // Start is called before the first frame update
+    void Start()
+    {
+        spawnerCordsX = gameObject.GetComponent<Transform>().position.x;
```
you can just use `transform.position.x` which is technically the same as `this.transform.position.x` monobehaviors already have a  variable that references its transform so you don't need to GetComponent

Kainkun commented on Assets/Scripts/InfiniteMana.cs:
```diff
@@ -0,0 +1,49 @@
+    {
+        if (Input.GetKeyDown("z"))
+        {
+            TakeDamage(100);
```
Not sure what this is

Kainkun commented on Assets/Scripts/ManaPickup.cs:
```diff
@@ -0,0 +1,23 @@
+    // Update is called once per frame
+    private void OnCollisionEnter2D(Collision2D collision)
+    {
+        if (collision.gameObject.name == "TempPlayer") //If an orb collides with whatever the object of the player's name is
```
using `if(GetComponent<PlatformerController>())` will be more efficient since it wont have to compare strings and its more reliable since the name "TempPlayer" will likely change

Kainkun reviewed (changes requested):
Hey, its working pretty well! There some tweaks id like for you to make. You can check out the comments in your code.

j-tomada committed:
Merge branch 'main' into InfiniteManaSpawner

j-tomada committed:
Made suggested changes

Changes to InfiniteMana:

Used Random instead of System Random

Orbs will now spawn on a random coordinate from a circle surrounding the spawner.

changed code to transform.position

Changes to ManaPickup:

Changed if statement to use GetComponent<PlatformerController>()

Kainkun reviewed (approved):
perfect!
